source/extensions/common/async_files/async_file_handle.h

```diff
@@ -0,0 +1,116 @@
+#pragma once
+
+#include <memory>
+#include <string>
+
+#include "source/common/buffer/buffer_impl.h"
+
+#include "absl/status/statusor.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace Common {
+namespace AsyncFiles {
+
+class AsyncFileAction;
+
+class AsyncFileContext : public std::enable_shared_from_this<AsyncFileContext> {
+public:
+  // Cancels any action that is currently queued. It is important to call abort if there is
+  // a request in flight whose on_complete callback's context is being destroyed.
+  //
+  // If the action is already being performed, abort cannot cancel the action, but will
+  // prevent calling of the callback.
+  //
+  // If the callback is currently being executed in a file worker thread, abort will block
+  // until file worker finishes the callback. (Callbacks should be extremely short-lived.)
+  // This is so that if the owner of the callback's context is being destroyed, we can
+  // guarantee that the context outlives any calls to it (e.g. we don't dispatch to a
+  // deleted event queue.)
+  virtual void abort() = 0;
+
+  // Action to create and open a temporary file.
+  // It is an error to createAnonymousFile on an AsyncFileContext that already has a file open.
+  //
+  // The path parameter is a path to a directory in which the anonymous file will be
+  // created (commonly "/tmp", for example). Even though an anonymous file is not linked and
+  // has no filename, the path can be important as it determines which physical hardware the file
+  // is written to (i.e. if you were to link() the file later, linking it to a path on a different
+  // device is an expensive operation; or you might prefer to write temporary files to a virtual
+  // filesystem or to a mounted disposable SSD.)
+  //
+  // There must not already be an action queued for this handle.
```

Comment:
Seems like this function should return an absl::Status with FailedPrecondition or something in this case? Same for other examples if this applies?